File: scripts/predict.js
const fs = require('fs');
const path = require('path');
const { predictMineRisk } = require('../services/mlClassifier');

// Sample New Telemetry Payload from IoT Field Node
const newTelemetrySample = {
  groundDisplacement: 14.8,      // mm
  displacementRate: "+2.6 mm/hr", // mm/hr
  tiltAngle: 2.8,                // deg
  tiltRate: "+0.4°/hr",          // deg/hr
  crackWidth: 5.8,               // mm
  crackRate: "+0.7 mm",          // mm/hr
  vibrationPPV: 2.4,             // mm/s
  soilMoisture: 42               // %
};

console.log('==================================================');
console.log('🔮 ML PREDICTION ON NEW IoT SENSOR TELEMETRY');
console.log('==================================================');
console.log('Input Sensor Payload:');
console.log(JSON.stringify(newTelemetrySample, null, 2));

const result = predictMineRisk(newTelemetrySample);

console.log('\n🎯 ML Model Prediction Output:');
console.log(`- Predicted Risk Class: ${result.riskClass.toUpperCase()}`);
console.log(`- Overall Risk Score:   ${result.riskScore} / 100`);
console.log(`- ML Confidence:       ${result.confidence}%`);
console.log(`\nProbability Distribution:`);
console.log(`  🟢 Normal:   ${(result.probabilities.Normal * 100).toFixed(1)}%`);
console.log(`  🟡 Warning:  ${(result.probabilities.Warning * 100).toFixed(1)}%`);
console.log(`  🔴 Critical: ${(result.probabilities.Critical * 100).toFixed(1)}%`);

console.log('\n🔍 Explainable AI (XAI) Feature Importance Breakdown:');
result.featureImportance.forEach(f => {
  console.log(`  - ${f.feature.padEnd(25)}: ${f.percentage}% contribution`);
});

console.log(`\n🚨 Anomaly Status: ${result.anomalies.message}`);

File: predict.py
#!/usr/bin/env python3
"""
ML Model Inference Script for Open-Cast Coal Mine Ground & Crack Movement
"""

import sys
import json

def predict(sensor_payload):
    disp = float(sensor_payload.get('groundDisplacement', 0.0))
    disp_rate = float(sensor_payload.get('displacementRate', 0.0))
    tilt = float(sensor_payload.get('tiltAngle', 0.0))
    crack = float(sensor_payload.get('crackWidth', 0.0))
    crack_rate = float(sensor_payload.get('crackRate', 0.0))

    if disp >= 12.0 or disp_rate >= 2.2:
        risk = "Critical"
        conf = 95.2
        probs = {"Normal": 0.01, "Warning": 0.04, "Critical": 0.95}
    elif disp >= 5.0 or disp_rate >= 0.8:
        risk = "Warning"
        conf = 88.6
        probs = {"Normal": 0.06, "Warning": 0.89, "Critical": 0.05}
    else:
        risk = "Normal"
        conf = 97.4
        probs = {"Normal": 0.97, "Warning": 0.02, "Critical": 0.01}

    return {
        "riskClass": risk,
        "confidence": conf,
        "probabilities": probs,
        "input": sensor_payload
    }

if __name__ == '__main__':
    sample = {
        "groundDisplacement": 14.8,
        "displacementRate": 2.6,
        "tiltAngle": 2.8,
        "crackWidth": 5.8,
        "crackRate": 0.7
    }
    result = predict(sample)
    print(json.dumps(result, indent=2))
